contracts/token/ERC6909/draft-ERC6909.sol

```diff
@@ -0,0 +1,123 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.20;
+
+import {IERC6909} from "../../interfaces/draft-IERC6909.sol";
+import {Context} from "../../utils/Context.sol";
+import {IERC165, ERC165} from "../../utils/introspection/ERC165.sol";
+
+contract ERC6909 is Context, ERC165, IERC6909 {
+    error ERC6909InsufficientBalance(address sender, uint256 balance, uint256 needed, uint256 id);
+    error ERC6909InsufficientAllowance(address spender, uint256 allowance, uint256 needed, uint256 id);
+    error ERC6909InvalidReceiver(address receiver);
+    error ERC6909InvalidSender(address sender);
+
+    mapping(uint256 id => mapping(address owner => uint256)) private _balances;
+
+    mapping(address owner => mapping(address operator => bool)) private _operatorApprovals;
+
+    mapping(address owner => mapping(address spender => mapping(uint256 id => uint256))) private _allowances;
+
+    function supportsInterface(bytes4 interfaceId) public view virtual override(ERC165, IERC165) returns (bool) {
+        return interfaceId == type(IERC6909).interfaceId || super.supportsInterface(interfaceId);
+    }
+
+    function balanceOf(address owner, uint256 id) public view virtual override returns (uint256) {
+        return _balances[id][owner];
+    }
+
+    function allowance(address owner, address spender, uint256 id) public view virtual override returns (uint256) {
+        return _allowances[owner][spender][id];
+    }
+
+    function isOperator(address owner, address spender) public view virtual override returns (bool) {
+        return _operatorApprovals[owner][spender];
+    }
+
+    function approve(address spender, uint256 id, uint256 amount) external virtual override returns (bool) {
+        address caller = _msgSender();
+        _allowances[caller][spender][id] = amount;
+
+        emit Approval(caller, spender, id, amount);
+        return true;
+    }
+
+    function setOperator(address spender, bool approved) external virtual override returns (bool) {
+        address caller = _msgSender();
+        _operatorApprovals[caller][spender] = approved;
+
+        emit OperatorSet(caller, spender, approved);
+        return true;
+    }
+
+    function transfer(address receiver, uint256 id, uint256 amount) external virtual override returns (bool) {
+        _update(_msgSender(), receiver, id, amount);
```

Comment:
I know this will make the `from != address(0)` check redundant but we're missing the `receiver != address(0)` check

```suggestion
        _transfer(_msgSender(), receiver, id, amount);
```